onnxruntime/core/providers/cpu/ml/tree_ensemble_common.h

```diff
@@ -201,103 +208,138 @@ Status TreeEnsembleCommon<InputType, ThresholdType, OutputType>::Init(
+  absl::flat_hash_map<TreeNodeElementId, uint32_t, TreeNodeElementId::hash_fn> idi;
+  idi.reserve(limit);
   max_feature_id_ = 0;
 
-  for (i = 0, limit = nodes_treeids.size(); i < limit; ++i) {
-    TreeNodeElement<ThresholdType>& node = nodes_[i];
-    node.id.tree_id = static_cast<int>(nodes_treeids[i]);
-    node.id.node_id = static_cast<int>(nodes_nodeids[i]);
+  for (i = 0; i < limit; ++i) {
+    TreeNodeElementId node_tree_id{static_cast<int>(nodes_treeids[i]),
+                                   static_cast<int>(nodes_nodeids[i])};
+    TreeNodeElement<ThresholdType> node;
     node.feature_id = static_cast<int>(nodes_featureids[i]);
     if (node.feature_id > max_feature_id_) {
       max_feature_id_ = node.feature_id;
     }
-    if (nodes_values_as_tensor.empty()) {
-      node.value = static_cast<ThresholdType>(nodes_values[i]);
-    } else {
-      node.value = nodes_values_as_tensor[i];
-    }
+    node.value_or_unique_weight = nodes_values_as_tensor.empty()
+                                      ? static_cast<ThresholdType>(nodes_values[i])
+                                      : nodes_values_as_tensor[i];
+
+    /* hitrates is not used for inference, they are ignored.
     if (nodes_hitrates_as_tensor.empty()) {
       node.hitrates = static_cast<ThresholdType>(i < nodes_hitrates.size() ? nodes_hitrates[i] : -1);
     } else {
       node.hitrates = i < nodes_hitrates_as_tensor.size() ? nodes_hitrates_as_tensor[i] : -1;
+    } */
+
+    node.flags = static_cast<uint8_t>(cmodes[i]);
+    node.truenode_inc_or_first_weight = 0;  // nodes_truenodeids[i] if not a leaf
+    node.falsenode_inc_or_n_weights = 0;    // nodes_falsenodeids[i] if not a leaf
+
+    if (i < static_cast<size_t>(nodes_missing_value_tracks_true.size()) && nodes_missing_value_tracks_true[i] == 1) {
+      node.flags |= static_cast<uint8_t>(MissingTrack::kTrue);
     }
-    node.mode = cmodes[i];
-    node.is_not_leaf = node.mode != NODE_MODE::LEAF;
-    node.truenode = nullptr;   // nodes_truenodeids[i];
-    node.falsenode = nullptr;  // nodes_falsenodeids[i];
-    node.missing_tracks = i < static_cast<size_t>(nodes_missing_value_tracks_true.size())
-                              ? (nodes_missing_value_tracks_true[i] == 1
-                                     ? MissingTrack::kTrue
-                                     : MissingTrack::kFalse)
-                              : MissingTrack::kNone;
-    node.is_missing_track_true = node.missing_tracks == MissingTrack::kTrue;
-    if (idi.find(node.id) != idi.end()) {
-      ORT_THROW("Node ", node.id.node_id, " in tree ", node.id.tree_id, " is already there.");
+    if (idi.find(node_tree_id) != idi.end()) {
+      ORT_THROW("Node ", node_tree_id.node_id, " in tree ", node_tree_id.tree_id, " is already there.");
     }
-    idi.insert(std::pair<TreeNodeElementId, TreeNodeElement<ThresholdType>*>(node.id, &node));
+    idi.insert(std::pair<TreeNodeElementId, uint32_t>(node_tree_id, i));
```

Comment:
Can we do:
```.cpp
 auto p = idi.insert(std::pair<TreeNodeElementId, uint32_t>(node_tree_id, i));
 if (!p.second) {
   ORT_THROW("Node ", node_tree_id.node_id, " in tree ", node_tree_id.tree_id, " is already there.");
 }
```

So we do not need `find()`

Reply:
done